### PR TITLE
Improve binary file check (fixes #205)

### LINKED ISSUE #205: Improve binary file check
Currently, we determine if a file is binary or not by [sniffing](https://github.com/editorconfig-checker/editorconfig-checker/blob/7d6a1d2ed4077afa7db777d4ef50dc1f2b9c620d/pkg/files/files.go#L143) the content type using the first 512 characters and [seeing](https://github.com/editorconfig-checker/editorconfig-checker/blob/a277484f00a0f5de082a1a39cded3bea95e29b92/pkg/config/config_test.go#L107) if it's `text/` or `application/octet-stream`. This produces lots of false positives, such as files that are UTF16.

I think a better solution would be to determine the [charset](https://pkg.go.dev/golang.org/x/net/html/charset) and then check for binary characters. I found that the tool [dos2unix](https://linux.die.net/man/1/dos2unix), [uses](https://sourceforge.net/p/dos2unix/dos2unix/ci/master/tree/dos2unix/common.c#l2034) the characters `\x00-\x08,\x0b,\x0e-\x1f` to determine if a file is binary. This is a very widely used tool, and this check seems reasonable. Thoughts?

### PR DESCRIPTION
**Pull request by rasa**
First draft. All tests pass. Let me know what you think.

### CODE REVIEW

**Review by mstruebing — COMMENTED** [review]
I'm getting some errors when running `make test` - I'm not sure it is related to my environment or not.

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1 @@
+/pkg/encoding/testdata/* -text
```
[the remark this answers is not in the document]
Comment (rasa) [reply .gitattributes:1]: It tells git not to change the line ending characters when checking out the files. So it effectively treats these files as binary, even though they are text files.

File: pkg/encoding/encoding.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		}
+		return contentString, charset, err
+	}
+	return decodedContentString, charset, nil
+}
+
+func detectText(contentBytes []byte) (string, error) {
+	detector := chardet.NewTextDetector()
+	// r, err := detector.DetectBest(contentBytes)
+	// return r.Charset, err
```
[the remark this answers is not in the document]
Comment (rasa) [reply pkg/encoding/encoding.go:120]: Ok

File: pkg/encoding/encoding.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	results, err := detector.DetectAll(contentBytes)
+	if err != nil {
+		return "", err
+	}
+	if len(results) == 0 {
+		return "", fmt.Errorf("Failed to determine charset")
+	}
+	confidence := -1
```
[the remark this answers is not in the document]
Comment (rasa) [reply pkg/encoding/encoding.go:129]: Yes, I can do some benchmarking. Unfortunately, except for html files (which often embed their charset in themselves) there is no way to know what encoding a file is without this sort of empirical logic. And go expects UTF-8 for strings. 

We might be able to speed things up by first scanning files to see if they contain binary characters, and then run the detect loop on only binary files to find UTF16s and UTF32s to convert them to UTF-8. But the line length check would probably miscount the length of lines on other non-UTF-8 files that are [multi-byte encodings](https://github.com/baulk/chardet/blob/master/multi_byte.go). 

We could also simply skip processing any files that aren't valid UTF-8 but that's an ugly "solution."

[the remark this answers is not in the document]
Comment (rasa) [reply pkg/encoding/encoding.go:129]: Also note that the current code reads each file twice ([here](https://github.com/editorconfig-checker/editorconfig-checker/blob/a6067702dd357ee10870df3b6734fd2e21a9045a/pkg/files/files.go#L128) and [here](https://github.com/editorconfig-checker/editorconfig-checker/blob/a6067702dd357ee10870df3b6734fd2e21a9045a/pkg/files/files.go#L143)).


This PR only reads each file once: https://github.com/editorconfig-checker/editorconfig-checker/blob/c1839f9fc4d46a4bd3ef78055258668183507634/pkg/validation/validation.go#L39-L60

**Comment by rasa** [comment]
I ran a quick test against https://github.com/torvalds/linux between main, and this PR:
```bash
$ time ec --no-color >ec-main.out

real    5m23.737s
user    0m0.000s
sys     0m0.000s

$ time ec --no-color >ec-pr206.out

real    7m14.026s
user    0m0.000s
sys     0m0.000s
```
And here's a diff of the results:
```diff
ross@indra:/c/github/torvalds$ diff -u -w --color=never ec-main.out ec-pr206.out
--- ec-main.out 2022-07-02 17:23:37.824216200 -0700
+++ ec-pr206.out 2022-07-02 17:30:51.863097200 -0700
@@ -28486,23 +28486,6 @@
        6: Trailing whitespace
 tools/perf/tests/attr/test-record-spe-physical-address:
        Wrong line endings or new final newline
-tools/perf/tests/pe-file.exe:
-       Wrong line endings or new final newline
-       Not all lines have the correct end of line character
-       30: Trailing whitespace
-       31: Trailing whitespace
-       32: Trailing whitespace
-tools/perf/tests/pe-file.exe.debug:
-       Wrong line endings or new final newline
-       Not all lines have the correct end of line character
-       149: Trailing whitespace
-       170: Trailing whitespace
-       171: Trailing whitespace
-       248: Trailing whitespace
-       269: Trailing whitespace
-       305: Trailing whitespace
-       306: Trailing whitespace
-       307: Trailing whitespace
 tools/perf/trace/beauty/beauty.h:
        69: Trailing whitespace
 tools/perf/trace/beauty/mmap_prot.sh:
@@ -28926,4 +28909,4 @@
 virt/kvm/kvm_main.c:
        2554: Trailing whitespace

-26105 errors found
+26090 errors found
```
per
```bash
$ file tools/perf/tests/pe-file.exe*
tools/perf/tests/pe-file.exe:       PE32+ executable (console) x86-64, for MS Windows
tools/perf/tests/pe-file.exe.debug: PE32+ executable (console) x86-64, for MS Windows
```

**Comment by rasa** [comment]
git's binary determination logic is at
https://github.com/git/git/blob/f9b95943b68b6b8ca5a6072f50a08411c6449b55/convert.c#L35

**Comment by mstruebing** [comment]
Thank you, I will try to get this in  this weekend!

**Comment by rasa** [comment]
The CI is [failing](https://github.com/editorconfig-checker/editorconfig-checker/runs/7373216789?check_suite_focus=true) due to the files not being found. All the tests pass when run locally, so I'm not sure what's up? Do they run fine locally for you too?

**Comment by mstruebing** [comment]
No I was assuming it is related to my machine/setup but it seems it is the same for CI

**Comment by rasa** [comment]
@mstruebing OK, lemme see if I can determine why they are failing, and fix it. On it now.

**Comment by rasa** [comment]
@mstruebing OK, I fixed the issues. The first is I hadn't committed `testdata/empty.txt` as it was in `.gitignore`. The second issue is the `TestGetRelativePath`[ deleted](https://github.com/editorconfig-checker/editorconfig-checker/blob/f81f2264ce7340a8cc4486655028ad5f48852be5/pkg/files/files_test.go#L127) the current working directory, so all disk-based tests [after that](https://github.com/editorconfig-checker/editorconfig-checker/blob/f81f2264ce7340a8cc4486655028ad5f48852be5/pkg/files/files_test.go#L173) would fail.

The fixes are in https://github.com/editorconfig-checker/editorconfig-checker/pull/210 . Thanks for merging, and sorry for the issues.

**Comment by mstruebing** [comment]
Thanks for taking care of this feature and especially the issues after merging, I really appreciate that 🙏

**Comment by rasa** [comment]
And for completeness, the dos2unix's binary check logic is [here](https://sourceforge.net/p/dos2unix/dos2unix/ci/master/tree/dos2unix/common.c#l2034)

**Pull request merged**